zkvm/src/signature/prover.rs

```diff
@@ -0,0 +1,191 @@
+        (
+            PartyAwaitingPrecommitments {
+                transcript,
+                multikey,
+                x_i,
+                r_i,
+                R_i,
+            },
+            precommitment,
+        )
+    }
+
+    pub fn receive_precommitments(
+        self,
+        nonce_precommitments: Vec<NoncePrecommitment>,
+    ) -> (PartyAwaitingCommitments, NonceCommitment) {
+        // Store received nonce precommitments in next state
+        (
+            PartyAwaitingCommitments {
+                transcript: self.transcript,
+                multikey: self.multikey,
+                x_i: self.x_i,
+                r_i: self.r_i,
+                nonce_precommitments,
+            },
+            self.R_i,
+        )
+    }
+}
+
+impl PartyAwaitingCommitments {
+    pub fn receive_commitments(
+        mut self,
+        nonce_commitments: Vec<NonceCommitment>,
+    ) -> Result<(PartyAwaitingSiglets, Siglet), VMError> {
+        // Check stored precommitments against received commitments
+        for (pre_comm, comm) in self
+            .nonce_precommitments
+            .iter()
+            .zip(nonce_commitments.iter())
+        {
+            // Make H(comm) = H(R_i)
+            let actual_precomm = comm.precommit();
+
+            // Compare H(comm) with pre_comm, they should be equal
+            // TBD: should we use ct_eq?
+            let equal = pre_comm.0.ct_eq(&actual_precomm.0);
+            if equal.unwrap_u8() == 0 {
+                return Err(VMError::InconsistentWitness);
+            }
+        }
+
+        // Make R = sum_i(R_i). nonce_commitments = R_i from all the parties.
+        let R: RistrettoPoint = nonce_commitments.iter().map(|R_i| R_i.0).sum();
+
+        // Make c = H(X_agg, R, m)
+        // The message should already have been fed into the transcript.
+        let c = {
+            self.transcript
+                .commit_point(b"X_agg", &self.multikey.aggregated_key().0);
+            self.transcript.commit_point(b"R", &R.compress());
+            self.transcript.challenge_scalar(b"c")
+        };
+
+        // Make a_i = H(L, X_i)
+        let X_i = VerificationKey((self.x_i.0 * RISTRETTO_BASEPOINT_POINT).compress());
+        let a_i = self.multikey.factor_for_key(&X_i);
+
+        // Generate siglet: s_i = r_i + c * a_i * x_i
+        let s_i = self.r_i.0 + c * a_i * self.x_i.0;
+
+        // Store received nonce commitments in next state
+        Ok((
+            PartyAwaitingSiglets {
+                multikey: self.multikey,
+                nonce_commitments,
+                c,
+            },
+            Siglet(s_i),
+        ))
+    }
+}
+
+impl PartyAwaitingSiglets {
+    pub fn receive_siglets(self, siglets: Vec<Siglet>) -> Signature {
+        // s = sum(siglets)
+        let s: Scalar = siglets.iter().map(|siglet| siglet.0).sum();
+        // R = sum(R_i). nonce_commitments = R_i
+        let R: RistrettoPoint = self.nonce_commitments.iter().map(|R_i| R_i.0).sum();
```

Comment:
we already compute it when doing F-S challenge, so we can pass it to the next state w/o recomputing it again.

Reply:
I don't understand... pass what to the next state? The `PartyAwaitingSiglets` is the last state, so it has to create a `Signature` (unless we want to get rid of the `PartyAwaitingSiglets` state and just have the aggregator add up the `s_i` and `R_i` components to make `Signature` themselves?

Reply:
sorry, i meant R is summed up first inside `PartyAwaitingCommitments::receive_commitments`, so we can pass it into `PartyAwaitingSiglets` and avoid recomputing it here.

Reply:
Ahh, that totally makes sense. Updated!